lib/cdo/url_converter.rb

```diff
@@ -0,0 +1,47 @@
+class UrlConverter
+  LEARN_CODE_ORG_REGEX = /\/\/learn.code.org(?=$|\/)/
+  HOUROFCODE_COM_REGEX = /\/\/hourofcode\.com(?=$|\/)/
+  CSEDWEEK_ORG_REGEX = /\/\/csedweek\.org(?=$|\/)/
+  DASHBOARD_REGEX = /\/\/studio.code.org(?=$|\/)/
+
+  # For reference, a 'host' is a domain and (optionally) port without a protocol.
+  # Examples: code.org, studio.code.org, localhost-studio.code.org:3000
+  # @see https://developer.mozilla.org/en-US/docs/Web/API/Location
+  def initialize(dashboard_host: nil, pegasus_host: nil, hourofcode_host: nil,
+    csedweek_host: nil)
+    @dashboard_host = dashboard_host
+    @pegasus_host = pegasus_host
+    @hourofcode_host = hourofcode_host
+    @csedweek_host = csedweek_host
+  end
+
+  # An 'origin' is a protocol, domain, and (optional) port.  This method may
+  # replace all three.
+  def replace_origin(url)
+    if @dashboard_host
+      raise 'Should not use learn.code.org' unless LEARN_CODE_ORG_REGEX.match(url).nil?
+    end
+
+    if @hourofcode_host && HOUROFCODE_COM_REGEX =~ url
+      url = url.gsub(HOUROFCODE_COM_REGEX, "//" + @hourofcode_host)
+    elsif @csedweek_host && CSEDWEEK_ORG_REGEX =~ url
+      url = url.gsub(CSEDWEEK_ORG_REGEX, "//" + @csedweek_host)
+    elsif @dashboard_host && DASHBOARD_REGEX =~ url
+      url = url.gsub(DASHBOARD_REGEX, "//" + @dashboard_host)
+    elsif @pegasus_host
+      # Handle i18n subdomains
+      if url =~ /\/\/\w+\.code\.org/
+        subdomain = /(?<=\/\/)\w+(?=\.code\.org)/.match(url)[0]
```

Comment:
Maybe a comment explaining what this sequence does?

Reply:
What? :scream: You don't find regexp intuitive and readable?

Reply:
I forgot string literal interpolation in regexes was a thing you could do. :+1: